flink-tests/src/test/java/org/apache/flink/test/state/operator/restore/AbstractOperatorRestoreTestBase.java

```diff
@@ -104,175 +91,97 @@ public static void beforeClass() {
+		String savepointPath = null;
 
 		// FLINK-6918: Retry cancel with savepoint message in case that StreamTasks were not running
 		// TODO: The retry logic should be removed once the StreamTask lifecycle has been fixed (see FLINK-4714)
-		boolean retry = true;
-		for (int i = 0; retry && i < 10; i++) {
-			Future<Object> future = jobManager.ask(msg, timeout);
-			result = Await.result(future, timeout);
-
-			if (result instanceof JobManagerMessages.CancellationFailure) {
-				Thread.sleep(50L);
-			} else {
-				retry = false;
+		while (deadline.hasTimeLeft() && savepointPath == null) {
+			try {
+				savepointPath = clusterClient.cancelWithSavepoint(
+					jobToMigrate.getJobID(),
+					targetDirectory.getAbsolutePath());
+			} catch (Exception e) {
+				if (!e.toString().matches(".* savepoint for the job .* failed.*")) {
+					throw e;
+				}
 			}
 		}
 
-		if (result instanceof JobManagerMessages.CancellationFailure) {
-			JobManagerMessages.CancellationFailure failure = (JobManagerMessages.CancellationFailure) result;
-			throw new Exception(failure.cause());
-		}
+		assertNotNull(savepointPath);
 
-		String savepointPath = ((JobManagerMessages.CancellationSuccess) result).savepointPath();
+		jobStatusFuture = FutureUtils.retrySuccesfulWithDelay(
+			() -> clusterClient.getJobStatus(jobToMigrate.getJobID()),
+			deadline.timeLeft().toMillis() / 50,
+			Time.milliseconds(50),
+			(jobStatus) -> jobStatus.equals(JobStatus.CANCELED),
+			TestingUtils.defaultScheduledExecutor());
 
-		// Wait until canceled
-		msg = new TestingJobManagerMessages.NotifyWhenJobStatus(jobToMigrate.getJobID(), JobStatus.CANCELED);
-		Await.ready(jobManager.ask(msg, timeout), timeout);
+		assertEquals(JobStatus.CANCELED, jobStatusFuture.get(deadline.timeLeft().toMillis(), TimeUnit.MILLISECONDS));
 
 		return savepointPath;
 	}
 
-	private void restoreJob(String savepointPath) throws Exception {
+	private void restoreJob(ClassLoader classLoader, ClusterClient<?> clusterClient, Deadline deadline, String savepointPath) throws Exception {
 		JobGraph jobToRestore = createJobGraph(ExecutionMode.RESTORE);
 		jobToRestore.setSavepointRestoreSettings(SavepointRestoreSettings.forPath(savepointPath, allowNonRestoredState));
 
-		Object msg;
-		Object result;
+		assertNotNull(jobToRestore.getJobID());
 
-		// Submit job graph
-		msg = new JobManagerMessages.SubmitJob(jobToRestore, ListeningBehaviour.DETACHED);
-		result = Await.result(jobManager.ask(msg, timeout), timeout);
+		clusterClient.submitJob(jobToRestore, classLoader);
 
-		if (result instanceof JobManagerMessages.JobResultFailure) {
-			JobManagerMessages.JobResultFailure failure = (JobManagerMessages.JobResultFailure) result;
-			throw new Exception(failure.cause());
-		}
-		Assert.assertSame(JobManagerMessages.JobSubmitSuccess.class, result.getClass());
+		CompletableFuture<JobStatus> jobStatusFuture =
+			clusterClient.getJobStatus(jobToRestore.getJobID());
+
+		while (deadline.hasTimeLeft()
```

Comment:
use `FutureUtils.retrySuccesfulWithDelay` instead?

Reply:
dammit, I had this fixed but didn't commit it...